### PR TITLE
feat: revamp (deposit/withdraw architecture)

### PR DESCRIPTION
**Pull request by ozwaldorf**
## Why?

Revamp the marketplace to solve atomicity issues

Adds deposit/withdraw of nfts and tokens, and bases the service flow around those functions

NFTs and unlocked funds can be withdrawn at any time, using their respective methods

## How?

- added nft_balance to balances struct, `hashmap<(collection, token id), principal>`
- added depositNFT/Fungible method
- added withdrawNFT/Fungible method

#### Listing
to create a listing that is available for direct buy, the nft must be deposited before the listing will go though.
```
call approve on crowns canister for marketplace on token
call depositNFT on marketplace canister
call makeListing with direct_buy = true
```
#### Buying
to direct buy a listing, you must deposit funds before the direct buy will go through
```
call approve on wicp for marketplace to access x tokens
call depositFungable on marketplace canister
call directBuy on marketplace canister for nft
```

### Offer Flow:
- a nft can be listed without a deposit if direct buy is disabled. A user can accept offers regardless of listing status (unlisted, listed for direct buy, listed for offers). When a seller decides to accept an offer, the nft must be deposited.
#### Making an offer
- to make an offer, you must deposit tokens first
- tokens are "locked" from withdraw until offer is denied or cancelled (both actions will attempt to auto withdraw, but if unsuccessful fall back to balance to manually withdraw)
- when an offer is made, the tokens are locked until the offer is cancelled
```
call approve on wicp for marketplace to access x tokens
call depositFungable on marketplace canister
call makeOffer for x amount of tokens
```
#### Accepting an offer
- to accept an offer, you must deposit nft at time of sale
- the offer amount is unlocked from the buyer and will attempt to send to the seller, if unsuccessful will fallback to balance that seller can manually withdraw
- after an offer is accepted, the offer is removed but the others remain (until denied or cancelled) and can still be accepted by the new owner
```
call approve on crowns canister for marketplace on token
call depositNFT on marketplace canister
call acceptOffer on marketplace canister
call capSync endpoint in kyasshu
```

### Withdraw methods
These are provided as a fallback way to withdraw a purchased nft or funds that failed to auto transfer

#### Fungible (wicp)
 - When you withdraw a fungible token, it will withdraw the amount a user has that is not locked
 - Users should only have a balance if a automatic withdraw failed, or they deposited more than they offered

#### NFT (crowns)
 - you can only withdraw a token that is not listed for direct buy

## Contribution checklist?

- [ ] The commit messages are detailed
- [ ] It does not break existing features (unless required)
- [ ] I have performed a self-review of my own code
- [ ] Documentation has been updated to reflect the changes
- [ ] Tests have been added or updated to reflect the changes
- [ ] All code formatting pass
- [ ] All lints pass
- [ ] All tests pass

## Security checklist?

- [ ] Injection has been prevented (parameterized queries, no eval or system calls)
- [ ] Sensitive data has been identified and is being protected properly

## Demo?
image

### CODE REVIEW

**Comment by ozwaldorf** [comment]
https://github.com/Psychedelic/nft-marketplace/actions/runs/2061992447

**Review by punkbit — DISMISSED** [review]
Mind add to, show this works by `.scripts/healthcheck.sh`, please?

File: marketplace/src/main.rs [hunk 1]
```diff
@@ -548,6 +548,81 @@ pub async fn get_all_offers(begin: u64, limit: u64) -> Vec<Offer> {
...
+        .get(&nft_canister_id)
+        .ok_or(MPApiError::NonExistentCollection)?;
+
+    if let Some(owner) = balances()
+        .nft_balances
+        .get_mut(&(nft_canister_id.clone(), token_id.clone()))
+    {
+        // marketplace owns this token, check if caller is the stored owner
+        if (caller == owner.clone()) {
+            *owner = self_id;
```
Comment (scott-dn) [inline marketplace/src/main.rs]: any reason to keep `self_id` as an owner even if after withdrawing succeed?

File: marketplace/src/main.rs [hunk 2]
```diff
@@ -548,6 +548,81 @@ pub async fn get_all_offers(begin: u64, limit: u64) -> Vec<Offer> {
...
+    .is_err()
+    {
+        return Err(MPApiError::TransferNonFungibleError);
+    }
+
+    // set token owner in balances ledger
+    balances()
+        .nft_balances
+        .entry((nft_canister_id, token_id))
+        .and_modify(|pid| *pid = caller.clone())
```
Comment (scott-dn) [inline marketplace/src/main.rs]: can remove clone as the caller is Principal type which is also implemented Copy trait.

**Review by scott-dn — APPROVED** [review]

**Pull request merged**